[Bug] Fix is_subnet_public to use main route table for subnets without explicit association

almightychang opened the pull request:
## Summary

- Fix `is_subnet_public()` to correctly identify public subnets that use the VPC's main route table
- When a subnet has no explicit route table association, find the main route table instead of using `route_tables[0]`
- Add explicit error handling when main route table cannot be found

## Root Cause

When a subnet has no explicit route table association, it implicitly uses the VPC's main route table. The previous implementation fetched all VPC route tables but only checked `route_tables[0]`, which may not be the main route table. This caused public subnets (using main route table with IGW) to be incorrectly identified as private.

## Changes

**`cli/src/pcluster/aws/ec2.py`**
- Find main route table by checking `Associations[].Main` flag
- Raise explicit exception if main route table not found

**`cli/tests/pcluster/aws/test_ec2.py`**
- Add helper functions for mocking route table responses
- Add test for subnet using main route table with IGW (public)
- Add test for subnet using main route table without IGW (private)

## Test plan

- [ ] Existing `test_is_subnet_public` passes
- [ ] New `test_is_subnet_public_with_main_route_table` passes
- [ ] New `test_is_subnet_public_main_route_table_no_igw` passes

Fixes #7173

almightychang committed:
[Bug] Fix is_subnet_public to use main route table for subnets without explicit association

When a subnet has no explicit route table association, it uses the VPC's
main route table. The previous implementation incorrectly used route_tables[0]
which may not be the main route table, causing public subnets to be
incorrectly identified as private.

This fix explicitly finds the main route table by checking the "Main" flag
in route table associations. If no main route table is found (which should
not happen in a valid VPC), an exception is raised with a helpful message.

Fixes #7173

himani2411 commented on cli/src/pcluster/aws/ec2.py:
```diff
@@ -572,6 +572,18 @@ def is_subnet_public(self, subnet_id):
 
+            # Find the main route table (subnets without explicit association use the main route table)
+            main_route_table = next(
+                (rt for rt in route_tables if any(assoc.get("Main") for assoc in rt.get("Associations", []))),
```
The simplest way would be to modify this line 
```
route_tables = self.describe_route_tables(filters=[{"Name": "vpc-id", "Values": [vpc_id]}])
```
to add another filter 
```
route_tables = self.describe_route_tables(filters=[
            {"Name": "vpc-id", "Values": [vpc_id]},
            {"Name": "association.main", "Values": ["true"]}
        ]
```

almightychang replied:
@himani2411 
I updated the lines as you suggested. Please check.

almightychang committed:
Use API filter association.main instead of Python filtering

Refactor is_subnet_public to use AWS API filter 'association.main=true'
to fetch only the main route table, instead of fetching all route tables
and filtering in Python. This is more efficient and cleaner.

Update test mocks to match the new API filter parameters.

himani2411 commented on cli/tests/pcluster/aws/test_ec2.py:
```diff
@@ -698,3 +727,55 @@ def test_is_subnet_public(boto3_stubber):
     assert AWSApi.instance().ec2.is_subnet_public(subnet_id) is True
+
+
+def test_is_subnet_public_with_main_route_table(boto3_stubber):
```
Can you combine these two separate function and keep the variables as pytest parameters using `@pytest.mark.parametrize`. You can see similar examples in the existing file

almightychang replied:
@himani2411
I missed the notifications. Completed.

almightychang committed:
Refactor main route table tests to use pytest.mark.parametrize

Combine test_is_subnet_public_with_main_route_table and
test_is_subnet_public_main_route_table_no_igw into a single
parameterized test for better maintainability.

himani2411 committed:
Merge branch 'develop' into fix/is-subnet-public-main-route-table
